features/storage/kvstore/kv_map/KVMap.h

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef _KV_MAP
+#define _KV_MAP
+
+#include "KVStore.h"
+#include "platform/PlatformMutex.h"
+#include "platform/SingletonPtr.h"
+#include "BlockDevice.h"
+#include "FileSystem.h"
+
+namespace mbed {
+
+#define  MAX_ATTACHED_KVS 3
+
+typedef struct {
+    KVStore *kvstore_main_instance;
+    KVStore *internal_store;
+    KVStore *external_store;
+    BlockDevice *internal_bd;
+    BlockDevice *external_bd;
+    FileSystem *external_fs;
+    uint32_t flags_mask;
+} kvstore_config_t;
+
+typedef struct {
+    char *partition_name;
+    kvstore_config_t *kv_config;
+} kv_map_entry_t;
+
+class KVMap : private mbed::NonCopyable<KVMap> {
+public:
+
+    /**
+     * @brief As a singleton, return the single instance of the class.
+     *        This class is a singleton for the following reasons:
+     *        - Ease of use, so you don't have to coordinate instantiations.
+     *        - Lazy instantiation of internal data, (which we can't achieve with simple static classes).
+     *
+     * @returns Singleton instance reference.
+     */
+    static KVMap &get_instance()
+    {
+        // Use this implementation of singleton (Meyer's) rather than the one that allocates
+        // the instance on the heap because it ensures destruction at program end (preventing warnings
+        // from memory checking tools, such as valgrind).
+        static KVMap instance;
+        return instance;
+    }
+
+    ~KVMap();
+
+    /**
+     * @brief Initializes KVMap
+     *
+     * @return 0 on success, negative error code on failure
+     */
+    int init();
+
+    /**
+     * @brief attach a KVStore partition configuration, and add it to the KVMap array
+     *
+     * @param partition_name string parameter contains the partition name.
+     * @param kv_config a configuration struct created by the kv_config or by the user.
+     * @return 0 on success, negative error code on failure
+     */
+    int attach(const char *partition_name, kvstore_config_t *kv_config);
```

Comment:
Is there a way to change a single kv_config option? For example, if I wanted to change the "external_fs" but keep the other config fields the same. Is there a way to get the kv_config structure that was defined previously?

Reply:
I think overriding a single parameter in the mbed_app.json will do that

Reply:
That doesn't work if what you want to overload is not one of the predefined options.

One option would be to allow NULL for configs you don't want to change.

It's fine if we don't want to change this, I'm just curious